### PR TITLE
ocaml 5.2.0

### PR DESCRIPTION
**Pull request by XVilka**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by XVilka** [comment]
Hmm, I am not sure, looks like camlp5 test is failing:
```
  ==> Testing camlp5
  ==> /opt/homebrew/Cellar/camlp5/8.03.00/bin/camlp5 /opt/homebrew/Cellar/camlp5/8.03.00/lib/ocaml/camlp5/pa_o.cmo /opt/homebrew/Cellar/camlp5/8.03.00/lib/ocaml/camlp5/pr_o.cmo /opt/homebrew/opt/ocaml/lib/ocaml/str/str.cma hi.ml
  Error while loading "/opt/homebrew/Cellar/camlp5/8.03.00/lib/ocaml/camlp5/pr_o.cmo": no implementation available for O_keywords.
  
  Error: camlp5: failed
  Error: camlp5: failed
  An exception occurred within a child process:
    Minitest::Assertion: Expected: 0
    Actual: 2
  /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/gems/minitest-5.23.1/lib/minitest/assertions.rb:183:in `assert'
  /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/gems/minitest-5.23.1/lib/minitest/assertions.rb:223:in `assert_equal'
  /opt/homebrew/Library/Homebrew/formula_assertions.rb:25:in `shell_output'
  /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/c/camlp5.rb:47:in `block in <class:Camlp5>'
  /opt/homebrew/Library/Homebrew/formula.rb:2786:in `block (3 levels) in run_test'
  /opt/homebrew/Library/Homebrew/extend/kernel.rb:529:in `with_env'
  /opt/homebrew/Library/Homebrew/formula.rb:2785:in `block (2 levels) in run_test'
  /opt/homebrew/Library/Homebrew/formula.rb:1189:in `with_logging'
  /opt/homebrew/Library/Homebrew/formula.rb:2784:in `block in run_test'
  /opt/homebrew/Library/Homebrew/mktemp.rb:75:in `block in run'
  /opt/homebrew/Library/Homebrew/mktemp.rb:75:in `chdir'
  /opt/homebrew/Library/Homebrew/mktemp.rb:75:in `run'
  /opt/homebrew/Library/Homebrew/formula.rb:3095:in `mktemp'
  /opt/homebrew/Library/Homebrew/formula.rb:2778:in `run_test'
  /opt/homebrew/Library/Homebrew/test.rb:46:in `block in <main>'
  /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.1/lib/ruby/3.3.0/timeout.rb:186:in `block in timeout'
  /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.1/lib/ruby/3.3.0/timeout.rb:41:in `handle_timeout'
  /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.1/lib/ruby/3.3.0/timeout.rb:195:in `timeout'
  /opt/homebrew/Library/Homebrew/test.rb:50:in `<main>'
```
This version is necessary for the ocaml-5.2.0 because it adds the support for it: https://github.com/camlp5/camlp5/commit/c2da7cea1f77ca7bdf2afb56ec27391d5715de2b

cc @chetmurthy

**Comment by XVilka** [comment]
> Hmm, I am not sure if this is a packaging issue, looks like camlp5 test is failing which might signal the mainstream bug:
> 
> ```
>   ==> Testing camlp5
>   ==> /opt/homebrew/Cellar/camlp5/8.03.00/bin/camlp5 /opt/homebrew/Cellar/camlp5/8.03.00/lib/ocaml/camlp5/pa_o.cmo /opt/homebrew/Cellar/camlp5/8.03.00/lib/ocaml/camlp5/pr_o.cmo /opt/homebrew/opt/ocaml/lib/ocaml/str/str.cma hi.ml
>   Error while loading "/opt/homebrew/Cellar/camlp5/8.03.00/lib/ocaml/camlp5/pr_o.cmo": no implementation available for O_keywords.
>   
>   Error: camlp5: failed
>   Error: camlp5: failed
>   An exception occurred within a child process:
>     Minitest::Assertion: Expected: 0
>     Actual: 2
>   /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/gems/minitest-5.23.1/lib/minitest/assertions.rb:183:in `assert'
>   /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/gems/minitest-5.23.1/lib/minitest/assertions.rb:223:in `assert_equal'
>   /opt/homebrew/Library/Homebrew/formula_assertions.rb:25:in `shell_output'
>   /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/c/camlp5.rb:47:in `block in <class:Camlp5>'
>   /opt/homebrew/Library/Homebrew/formula.rb:2786:in `block (3 levels) in run_test'
>   /opt/homebrew/Library/Homebrew/extend/kernel.rb:529:in `with_env'
>   /opt/homebrew/Library/Homebrew/formula.rb:2785:in `block (2 levels) in run_test'
>   /opt/homebrew/Library/Homebrew/formula.rb:1189:in `with_logging'
>   /opt/homebrew/Library/Homebrew/formula.rb:2784:in `block in run_test'
>   /opt/homebrew/Library/Homebrew/mktemp.rb:75:in `block in run'
>   /opt/homebrew/Library/Homebrew/mktemp.rb:75:in `chdir'
>   /opt/homebrew/Library/Homebrew/mktemp.rb:75:in `run'
>   /opt/homebrew/Library/Homebrew/formula.rb:3095:in `mktemp'
>   /opt/homebrew/Library/Homebrew/formula.rb:2778:in `run_test'
>   /opt/homebrew/Library/Homebrew/test.rb:46:in `block in <main>'
>   /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.1/lib/ruby/3.3.0/timeout.rb:186:in `block in timeout'
>   /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.1/lib/ruby/3.3.0/timeout.rb:41:in `handle_timeout'
>   /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.1/lib/ruby/3.3.0/timeout.rb:195:in `timeout'
>   /opt/homebrew/Library/Homebrew/test.rb:50:in `<main>'
> ```
> 
> This version is necessary for the ocaml-5.2.0 because it adds the support for it: [camlp5/camlp5@c2da7ce](https://github.com/camlp5/camlp5/commit/c2da7cea1f77ca7bdf2afb56ec27391d5715de2b)
> 
> cc @chetmurthy

Nevermind, I see it is likely because of these changes:
- https://github.com/camlp5/camlp5/commit/9b1c59be416338bfbe16b248cd5e940929abb004
- https://github.com/camlp5/camlp5/commit/f5f17323f2abc082bb808dedaec2299b874042ed

**Comment by XVilka** [comment]
No, using the new `pr_op` module didn't help, just produced a new error:
```
> /opt/homebrew/Cellar/camlp5/8.03.00/bin/camlp5 /opt/homebrew/Cellar/camlp5/8.03.00/lib/ocaml/camlp5/pa_o.cmo /opt/homebrew/Cellar/camlp5/8.03.00/lib/ocaml/camlp5/pr_op.cmo /opt/homebrew/opt/ocaml/lib/ocaml/str/str.cma hi.ml
  Error while loading "/opt/homebrew/Cellar/camlp5/8.03.00/lib/ocaml/camlp5/pr_op.cmo": execution of module initializers in the shared library failed: Failure("level top not found").
```

**Comment by chetmurthy** [comment]
I'm not sure where the test is in the brew build process, but the problem is that where you load "pr_o.cmo", now you must load "o_keywords.cmo pr_o.cmo".  But the old way of invoking camlp5 for preprocessing (e.g. "camlp5 pa_o.cmo pr_o.cmo") is cumbersome and predates findlib.  Instead you can use "not-ocamlfind preprocess -syntax camlp5o -package camlp5.pa_o,camlp5.pr_o") as is done here: https://github.com/drjdn/p5scm/pull/4/files

**Comment by chetmurthy** [comment]
Yes, "pr_op.cmo" provides extensions to the "pr_o.cmo" printer.  The entire methodology of loading individual .cmo files into the camlp5 process is flawed, b/c you have to specify all the files you want, and in a compatible load-order.  There is no way to get inference of which other files are needed, for the files you specify, etc.  All of this sort of thing is the proper domain of findlib, which knows exactly these things and does it when constructing preprocessor commands -- so that is what `not-ocamlfind preprocess` hooks into.

**Comment by XVilka** [comment]
> Yes, "pr_op.cmo" provides extensions to the "pr_o.cmo" printer. The entire methodology of loading individual .cmo files into the camlp5 process is flawed, b/c you have to specify all the files you want, and in a compatible load-order. There is no way to get inference of which other files are needed, for the files you specify, etc. All of this sort of thing is the proper domain of findlib, which knows exactly these things and does it when constructing preprocessor commands -- so that is what `not-ocamlfind preprocess` hooks into.

Thanks for pointing this. This is my first time updating Homebrew package, so I don't know why this particular test was written this way, but from the comment purpose it was made to check the binary compatibility with the OCaml compiler. I will try to change to that `not-ocamlfind` instead.

**Comment by chetmurthy** [comment]
Ah, you're trying to check that the output of `camlp5` is binary-compatible with the ocaml compiler ?  In that case, you shouldn't be loading "pr_o.cmo" -- that produces text source.  If you want to check binary compatibility, you could use `not-ocamlfind preprocess -syntax camlp5o -package camlp5.pa_o,camlp5.pr_dump`) .  This would output the result of preprocessing as an OCaml binary AST.

File: Formula/c/camlp5.rb [hunk 1]
```diff
@@ -51,7 +50,7 @@ def install
       # ocaml files are in sync with the camlp5 files.  If camlp5 has been
       # compiled with an older version of the ocaml compiler, then an error
       # "interface mismatch" will occur.
-      shell_output("#{bin}/camlp5 #{lib}/ocaml/camlp5/pa_o.cmo #{lib}/ocaml/camlp5/pr_o.cmo " \
+      shell_output("#{bin}/camlp5 #{lib}/ocaml/camlp5/pa_o.cmo #{lib}/ocaml/camlp5/o_keywords.cmo #{lib}/ocaml/camlp5/pr_o.cmo " \
                    "#{ocaml.opt_lib}/ocaml/str/str.cma hi.ml")
```
Comment (carlocab) [inline Formula/c/camlp5.rb:54]: Let's try this:
```suggestion
      system bin/"camlp5", lib/"ocaml/camlp5/pa_o.cmo",
                           lib/"ocaml/camlp5/o_keywords.cmo",
                           lib/"ocaml/camlp5/pr_o.cmo",
                           ocaml.opt_lib/"ocaml/str/str.cma hi.ml"
```

Comment (carlocab) [inline Formula/c/camlp5.rb:54]:
```suggestion
      shell_output("#{bin}/camlp5 #{lib}/ocaml/camlp5/pa_o.cmo " \
                   "#{lib}/ocaml/camlp5/o_keywords.cmo " \
                   "#{lib}/ocaml/camlp5/pr_o.cmo " \
                   "#{ocaml.opt_lib}/ocaml/str/str.cma hi.ml")
```

**Comment by chetmurthy** [comment]
If the issue here is tests, you could use the *extensive* test-suite that comes with Camlp5.  When I took over maintenance, the test-suite was pretty limited, but I've added a massive set of tests of all sorts of function.  You can look at the `opam` file to see how they're invoked.  And if you have other tests in mind, I can add those too!

**Comment by XVilka** [comment]
> Ah, you're trying to check that the output of `camlp5` is binary-compatible with the ocaml compiler ? In that case, you shouldn't be loading "pr_o.cmo" -- that produces text source. If you want to check binary compatibility, you could use `not-ocamlfind preprocess -syntax camlp5o -package camlp5.pa_o,camlp5.pr_dump`) . This would output the result of preprocessing as an OCaml binary AST.

For now I use the old solution, since the `not-ocamlfind` is not available at this stage.

**Review by carlocab — APPROVED** [review]
Thanks!

**Comment by XVilka** [comment]
Strange error while building Coq:
```
  ==> ./configure -prefix /home/linuxbrew/.linuxbrew/Cellar/coq/8.19.1_1 -mandir /home/linuxbrew/.linuxbrew/Cellar/coq/8.19.1_1/share/man -libdir /home/linuxbrew/.linuxbrew/lib/ocaml/coq -docdir /home/linuxbrew/.linuxbrew/Cellar/coq/8.19.1_1/share/coq/latex
  ocamlfind: Config file not found - neither /home/linuxbrew/.linuxbrew/etc/findlib.conf nor the directory /home/linuxbrew/.linuxbrew/etc/findlib.conf.d
  Error while running '/home/linuxbrew/.linuxbrew/opt/ocaml-findlib/bin/ocamlfind query findlib -format %v' (exit code 2)
  Configuration script failed!
```

**Comment by XVilka** [comment]
> Strange error while building Coq:
> 
> ```
>   ==> ./configure -prefix /home/linuxbrew/.linuxbrew/Cellar/coq/8.19.1_1 -mandir /home/linuxbrew/.linuxbrew/Cellar/coq/8.19.1_1/share/man -libdir /home/linuxbrew/.linuxbrew/lib/ocaml/coq -docdir /home/linuxbrew/.linuxbrew/Cellar/coq/8.19.1_1/share/coq/latex
>   ocamlfind: Config file not found - neither /home/linuxbrew/.linuxbrew/etc/findlib.conf nor the directory /home/linuxbrew/.linuxbrew/etc/findlib.conf.d
>   Error while running '/home/linuxbrew/.linuxbrew/opt/ocaml-findlib/bin/ocamlfind query findlib -format %v' (exit code 2)
>   Configuration script failed!
> ```

This looks similar to https://github.com/Homebrew/homebrew-test-bot/issues/805 https://github.com/Homebrew/homebrew-core/pull/154837/commits/4f236f226fbd7f04c0bc371ba2344a70817cec18, just different paths.

**Comment by XVilka** [comment]
@carlocab I fixed the Coq installation/testing problem, could you please restart the macOS 13 x86_64? I think it's unrelated problem:
```
 /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.2/bin/bundle clean
  ==> Testing hyperkit
  ==> /usr/local/Cellar/hyperkit/0.20210107_1/bin/hyperkit -v 2>&1
  /usr/bin/env tar --extract --no-same-owner --file /Users/brew/Library/Caches/Homebrew/downloads/840a6f58d851f8ac4dbfee434980e81b172cdb88593b1f5bfacf1fb074f981c5--tinycorelinux_8.x.tar.gz --directory /private/tmp/homebrew-unpack20240604-96003-ogq8uu
  /usr/bin/env cp -pR /private/tmp/homebrew-unpack20240604-96003-ogq8uu/tinycorelinux_8.x/. /private/tmp/hyperkit--tinycorelinux-20240604-96003-mm92c8/tinycorelinux_8.x
  ==> expect test_hyperkit.exp
  spawn /usr/local/Cellar/hyperkit/0.20210107_1/bin/hyperkit -A -m 512M -s 0:0,hostbridge -s 31,lpc -l com1,stdio -f kexec,./vmlinuz,./initrd.gz,earlyprintk=serial console=ttyS0
  
  Using fd 8 for I/O notifications
  
  send: spawn id exp6 not open
      while executing
  "send "sudo halt\r\n""
      (file "test_hyperkit.exp" line 17)
  
  Error: hyperkit: failed
  Error: hyperkit: failed
  An exception occurred within a child process:
    BuildError: Failed executing: expect test_hyperkit.exp
```
As for the PR, it should be ready.

**Review by carlocab — APPROVED** [review]

**Comment by XVilka** [comment]
No, it didn't help. But I have absolutely zero idea what could be the cause of this problem and why it happens only on the macOS 13 x86_64

**Comment by carlocab** [comment]
`hyperkit` has been broken for a while -- let's see how the other tests go. If everything else looks good we can merge this.

**Review by carlocab — APPROVED** [review]
Test failures other than `hyperkit` look unrelated.

**Review by chenrui333 — APPROVED** [review]

**Comment by chenrui333** [comment]
@XVilka, thanks for your contribution to Homebrew! :tada: 🥇

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!

File: Formula/h/hyperkit.rb [hunk 2]
```diff
@@ -4,12 +4,10 @@ class Hyperkit < Formula
   url "https://github.com/moby/hyperkit/archive/refs/tags/v0.20210107.tar.gz"
   sha256 "095f5f5ef550d7cad10e4d13e9c9ce8b58cc319d654a6d837d8d87ee70537835"
   license "BSD-2-Clause"
+  revision 1
 
   bottle do
-    sha256 cellar: :any_skip_relocation, ventura:  "3b67078315551718bc3c752b943b933713ddb69058f3cb72a0f65faa6e9295ab"
-    sha256 cellar: :any_skip_relocation, monterey: "da3b0d0374a85af5c649c86fb7796c1eecae468f5783bbb994a96d807e60712a"
-    sha256 cellar: :any_skip_relocation, big_sur:  "f96e7270e9e853ce33f2195136b11338a5cf4d612ee50f3dd51b5c8506b4efcb"
-    sha256 cellar: :any_skip_relocation, catalina: "cd58afe172473278d3ed9404e9d25e10bee487fb4e27cd6de39c950a0ccaca87"
```
Comment (chenrui333) [inline Formula/h/hyperkit.rb:12]: need to followup on this

Comment (chenrui333) [reply Formula/h/hyperkit.rb:12]: https://github.com/Homebrew/homebrew-core/actions/runs/9394971412

Comment (XVilka) [reply Formula/h/hyperkit.rb:12]: Hyperkit really looks like abandonware now: 
- https://github.com/moby/hyperkit/commits/master/
- https://github.com/moby/hyperkit/issues
- https://github.com/moby/hyperkit/pulls 

Not much could probably be done for it, until developers/maintainers step in.

Comment (XVilka) [reply Formula/h/hyperkit.rb:12]: Apologize for tagging you @avsm, but since you were hyperkit developer, you might be able to help reviving it.

Comment (chenrui333) [reply Formula/h/hyperkit.rb:12]: yeah, there are no new commits for the past two years.

Comment (avsm) [reply Formula/h/hyperkit.rb:12]: It's not so much abandonware, but stable. There's no need to keep it up-to-date with the very latest OCaml, and I doubt there are many users accessing it via Homebrew. I'm happy to it to be removed from Homebrew to ensure there's no blocker on upstream OCaml updates. Copying in @djs55 as another maintainer of Hyperkit to confirm.

Comment (djs55) [reply Formula/h/hyperkit.rb:12]: I would describe it as stable -- it's still shipped in Docker Desktop for Intel Macs. Removing hyperkit from home-brew is fine.

**Pull request merged**